tests/phpunit/CRM/Membershipextrasimporterapi/EntityImporter/MembershipTest.php

```diff
@@ -0,0 +1,306 @@
+
+/**
+ *
+ * @group headless
+ */
+class CRM_Membershipextrasimporterapi_EntityImporter_MembershipTest extends BaseHeadlessTest {
+
+  private $sampleRowData = [
+    'membership_external_id' => 'test1',
+    'membership_type' => 'Student',
+    'membership_join_date' => '20180101000000',
+    'membership_start_date' => '20190101000000',
+    'membership_end_date' => '20191231000000',
+    'membership_status' => 'Current',
+    'membership_is_status_overridden' => 0,
+  ];
+
+  private $contactId;
+
+  private $recurContributionId;
+
+  public function setUp() {
+    $this->contactId = ContactFabricator::fabricate()['id'];
+
+    $recurContributionParams = ['contact_id' => $this->contactId, 'amount' => 50, 'frequency_interval' => 1];
+    $this->recurContributionId = RecurContributionFabricator::fabricate($recurContributionParams)['id'];
+
+    MembershipTypeFabricator::fabricate(['name' => 'Student', 'minimum_fee' => 50]);
+  }
+
+  public function testImportNewMembership() {
+    $beforeImportIds = $this->getMembershipsByContactId($this->contactId);
+
+    $membershipImporter = new MembershipImporter($this->sampleRowData, $this->contactId, $this->recurContributionId);
+    $newMembershipId = $membershipImporter->import();
+
+    $afterImportIds = $this->getMembershipsByContactId($this->contactId);
+
+    $importSucceed = FALSE;
+    if (empty($beforeImportIds) && $afterImportIds[0] == $newMembershipId) {
+      $importSucceed = TRUE;
+    }
+
+    $this->assertTrue($importSucceed);
+  }
+
+  public function testImportExistingMembershipWillNotCreateNewOne() {
+    $this->sampleRowData['payment_plan_external_id'] = 'test2';
+
+    $firstImport = new MembershipImporter($this->sampleRowData, $this->contactId, $this->recurContributionId);
+    $firstMembershipId = $firstImport->import();
+
+    $secondImport = new MembershipImporter($this->sampleRowData, $this->contactId, $this->recurContributionId);
+    $secondMembershipId = $secondImport->import();
+
+    $this->assertEquals($firstMembershipId, $secondMembershipId);
+  }
+
+  public function testImportWillSetCorrectRecurContributionId() {
+    $this->sampleRowData['membership_external_id'] = 'test3';
+
+    $membershipImporter = new MembershipImporter($this->sampleRowData, $this->contactId, $this->recurContributionId);
+    $newMembershipId = $membershipImporter->import();
+
+    $newMembership = $this->getMembershipById($newMembershipId);
+
+    $this->assertEquals($this->recurContributionId, $newMembership['contribution_recur_id']);
+  }
+
+  public function testImportWillCreateExternalIdCustomFieldRecord() {
+    $this->sampleRowData['membership_external_id'] = 'test4';
+
+    $membershipImporter = new MembershipImporter($this->sampleRowData, $this->contactId, $this->recurContributionId);
+    $newMembershipId = $membershipImporter->import();
+
+    $sqlQuery = "SELECT entity_id as id FROM civicrm_value_membership_ext_id WHERE external_id = %1 AND entity_id = {$newMembershipId}";
```

Comment:
Similar to the above comment, you are mixing (curly) syntax with sql params

Reply:
same as above, $newMembershipId is not user provided value unlike external_id, so I think it is fine.